File: leetcode_alg/solution11.cpp
//
// Created by wangheng on 2021/6/17.
// 盛水问题: 双指针算法
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int maxArea(vector<int> &height) {
        int l = 0;
        int r = height.size()-1;
        int maxArea=0;
        while(l<r){
            int area = min(height[l], height[r]) * (r-l);
            if(maxArea< area){
                maxArea = area;
            }
            if(height[l] < height[r]){
                l++;
            }else{
                r--;
            }
        }
        return maxArea;
    }
};

int main() {

    return 0;
}

File: golang/solution/solution11.go
package solution

import (
	"fmt"
	"math"
)

// Container with most water

func maxArea(height []int) int {

	l, r := 0, len(height)-1
	var maxRes int = 0

	for l < r {
		area := math.Min(float64(height[l]), float64(height[r])) * float64(r-l)
		if area > float64(maxRes) {
			maxRes = int(area)
		}
		if height[l] <= height[r] {
			l++
		} else {
			r--
		}
	}
	return maxRes
}

func Solution11() {
	var height [5]int = [5]int{3, 4, 2, 6, 7}
	fmt.Println(maxArea(height[:]))
}
